jax/interpreters/xla.py

```diff
@@ -122,11 +122,14 @@ def array_result_handler(device: Optional[Device], aval: core.ShapedArray):
 }
 
 def device_put(x, device: Optional[Device] = None) -> Tuple[Any]:
-  x = canonicalize_dtype(x)
-  try:
-    return device_put_handlers[type(x)](x, device)
-  except KeyError as err:
-    raise TypeError(f"No device_put handler for type: {type(x)}") from err
+  handler = device_put_handlers.get(type(x))
+  if handler:
+    x = canonicalize_dtype(x)
+    return handler(x, device)
+  elif hasattr(x, '__jax_array__'):
+    return device_put(x.__jax_array__(), device)
```

Comment:
Maybe overkill, but we might add a guardrail here against infinite recursion here and below.

Reply:
Hm, what kind of guard do you have in mind?

That sounds like it'd just be a bug in the `__jax_array__` implementation, and trying to guard against a potential infinite recursion just to raise a different error doesn't seem clearly better than just letting people get the infinite recursion error.